[Merged by Bors] - feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

strihanje01 opened the pull request:
Prove the noncommutative Freiman-Kneser theorem for doubling less than 2 - ε

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

strihanje01 committed:
First commit

strihanje01 committed:
Second commit

strihanje01 committed:
Third commit

strihanje01 committed:
Fourth commit

strihanje01 committed:
Fifth commit

strihanje01 committed:
Sixth commit

strihanje01 committed:
Seventh commit

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

add noncommutative Freiman-Kneser theorem for doubling less than 2-\epsilon

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

add Freiman-Kneser theorem for doubling less than 2-\epsilon

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

add noncommutative Freiman-Kneser theorem for doubling less than 2-\epsilon

strihanje01 commented:
@YaelDillies Here is the theorem for doubling smaller than $2-\epsilon$

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

rewrite expansion, connectivity, fraction and atom and adapt the proof for doubling less than `2 - \epsilon`

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

rewrite expansion, connectivity, fraction and atom and adapt the proof for doubling less than `2 - \epsilon`

strihanje01 commented:
@YaelDillies So, about min_imports, all those imports are what I get when I run it... Now, I've been having problems with my commits lately (here and in Kneser's theorem PR), they are failing checks (after ~50min, seemingly in the same way) and I can't locate the issue. Could you help me?

strihanje01 committed:
Merge remote-tracking branch 'upstream/master' into strihanje01_hamidounesFreimanKneserTheorem

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -328,4 +333,407 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+  by_contra h
+  simp only [Disjoint, le_eq_subset, bot_eq_empty,
+              subset_empty, not_forall, ← nonempty_iff_ne_empty] at h
+  obtain ⟨B, hBx, ⟨hBy, ⟨b, hb⟩⟩⟩ := h
```
Pointy bracket right-associate by default:
```suggestion
  obtain ⟨B, hBx, hBy, b, hb⟩ := h
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -328,4 +333,407 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    BddBelow (Set.range fun A : {A : Finset G // A.Nonempty} ↦ expansion K S A) :=
+  ⟨0, by simp [lowerBounds, *]⟩
+
+private noncomputable def connectivity (K : ℝ) (S : Finset G) : ℝ :=
```
Can you provide some quick docstrings for all the `private def`s?

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -328,4 +333,407 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+      simpa only [← IsAtom.eq_of_inter_nonempty hK.le hS (IsAtom.smul_finset n⁻¹ hN)
+        (IsAtom.smul_finset a⁻¹ (IsAtom.smul_finset n⁻¹ hN)) this] using self_mem_smul_carrier a⁻¹
+  }
+  refine ⟨H, Fintype.ofFinset (n⁻¹ •> N) (fun a => by
+      simpa only [← mem_coe, coe_smul_finset] using H.mem_carrier),
+    by simpa only [Subgroup.coe_set_mk, Set.toFinset_smul_set, toFinset_coe, H]
+      using IsAtom.smul_finset n⁻¹ hN
+  ⟩
```
```suggestion
  refine ⟨H, Fintype.ofFinset (n⁻¹ •> N) fun a => ?_, ?_⟩
  · simpa only [← mem_coe, coe_smul_finset] using H.mem_carrier
  · simpa only [Subgroup.coe_set_mk, Set.toFinset_smul_set, toFinset_coe, H]
      using IsAtom.smul_finset n⁻¹ hN
```
is much more readable, and takes less space too!

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -328,4 +333,407 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+`S` is covered by at most `2 / ε - 1` right cosets of `H`.
+In particular, for `A = S`, we get a characterisation of sets of doubling less than `2 - ε`. -/
+theorem doubling_lt_two {ε : ℝ} (hε₀ : 0 < ε) (hε₁ : ε ≤ 1) (hS : S.Nonempty)
+    (hA : ∃ (A : Finset G) (_ : #A ≥ #S), #(A * S) ≤ (2 - ε) * #S) :
+    ∃ (H : Subgroup G) (_ : Fintype H) (_ : Fintype.card H ≤ (2 / ε - 1) * #S) (Z : Finset G)
+      (_ : #Z ≤ 2 / ε - 1), (S : Set G) ⊆ (H : Set G) * Z := by
```
```suggestion
    (hA : ∃ A : Finset G, #S ≤ #A ∧ #(A * S) ≤ (2 - ε) * #S) :
    ∃ (H : Subgroup G) (_ : Fintype H)(Z : Finset G),
      Fintype.card H ≤ (2 / ε - 1) * #S ∧ #Z ≤ 2 / ε - 1 ∧ (S : Set G) ⊆ H * Z := by
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -328,4 +333,407 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+  have calc₁ : ex (Set.toFinset H) ≤ (1 - ε / 2) * #S := by
+    calc
+          ex (Set.toFinset H)
+      _ = κ                               := IsAtom.isFragment hH
```
```suggestion
      _ = κ                               := hH.isFragment
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -328,4 +333,407 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+      _ ≤ (2 / ε - 1) * #(H : Set G).toFinset / #(H : Set G).toFinset := ?_
+      _ = 2 / ε - 1                                                   := by field_simp; ring
+    gcongr
+    -- Finally, to show `#(HS) ≤ (2 / ε - 1) * #H`, we multiply bothe sides by `1 - K = ε / 2` and
```
```suggestion
    -- Finally, to show `#(HS) ≤ (2 / ε - 1) * #H`, we multiply both sides by `1 - K = ε / 2` and
```

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

add noncommutative Freiman-Kneser theorem for doubling less than `2-\epsilon`

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

resolve the first round of comments; add docstrings for expansion, connectivity, fraction, atom

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into strihanje01_hamidounesFreimanKneserTheorem

YaelDillies committed:
golf

YaelDillies committed:
shake

strihanje01 committed:
Merge branch 'master' into strihanje01_hamidounesFreimanKneserTheorem

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

resolve merge confilcts

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

resolve min imports

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

resolve new round of comments

strihanje01 commented:
-awaiting-author

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -342,7 +342,6 @@ private lemma op_smul_eq_iff_mem {H : Subgroup G} {c : Set G} {x : G}
   change _ = _ <• _
   rw [eq_comm, smul_eq_iff_eq_inv_smul, ← op_inv, op_smul_op_smul, rightCoset_mem_rightCoset]
   rwa [← op_smul_eq_mul, op_inv, ← SetLike.mem_coe, ← Set.mem_smul_set_iff_inv_smul_mem]
-
```
Can you revert this accident?

YaelDillies committed:
small golfs

YaelDillies committed:
more golfs

YaelDillies reviewed (approved):
I've golfed a few things here and there and it looks good to me now. Thank you for your patience!

maintainer merge

b-mehta reviewed (commented):
Thanks! I've left one comment.

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): Hamidoune's Freiman-Kneser theorem for nonabelian groups

add corollary for A=S

strihanje01 committed:
Merge branch 'master' into strihanje01_hamidounesFreimanKneserTheorem

b-mehta commented:
Thanks!

bors merge
